app/bundles/LeadBundle/Controller/LeadController.php

```diff
@@ -764,6 +764,171 @@ private function uploadAvatar(Lead $lead)
+
+                    //pull the data from the form in order to apply the form's formatting
+                    foreach ($form as $f) {
+                        $name = $f->getName();
+                        if (strpos($name, 'field_') === 0) {
+                            $data[$name] = $f->getData();
+                        }
+                        
+
+                        $secLeadId = $data['lead_to_merge'];
+                        $secLead =  $model->getRepository('MauticLeadBundle:Lead')->getLeadsByUniqueFields(array(),$secLeadId)[0];
+                        //$secLead = $model->getEntity($secLeadId);
+
+                        if ($secLead === null) {
+            				return $this->postActionRedirect(
+                				array_merge(
+                    				$postActionVars,
+                    				array(
+                        				'flashes' => array(
+                            				array(
+                                				'type'    => 'error',
+                                				'msg'     => 'mautic.lead.lead.error.notfound',
+                                				'msgVars' => array('%id%' => $secLead->getId())
+                            				)
+                        				)
+                    				)
+                				)
+            				);
+        				} elseif (!$this->factory->getSecurity()->hasEntityAccess(
+            				'lead:leads:editown',
+            				'lead:leads:editother',
+            				$mainLead->getOwner())
+        				) {
+            				return $this->accessDenied();
+        				} elseif ($model->isLocked($secLead)) {
+            				//deny access if the entity is locked
+            				return $this->isLocked($postActionVars, $secLead, 'lead.lead');
+        				}
+
+                        if ($mainLead->getDateAdded() > $secLead->getDateAdded()){
+                            list($secLead,$mainLead) = array($mainLead,$secLead);
+                        }
+
+        				//Both leads are good so now we merge them
+                        $mainLead = $model->mergeLeads($mainLead, $secLead);
+                        $model->saveEntity($mainLead);
+                    }       
+                }
+            } 
+
+         	$route          = 'mautic_lead_action';  
+       		$func           = 'view';
+				
+			$viewParameters = array(
+					'objectId'     => $mainLead->getId(),
+                    'objectAction' => 'view',
+            );
+
+            return $this->postActionRedirect(
+                array(
+                    'returnUrl'       => $this->generateUrl($route, $viewParameters),
+                    'viewParameters'  => $viewParameters,
+                    'contentTemplate' => 'MauticLeadBundle:Lead:'.$func,
+                    'passthroughVars' => array(
+                        'mauticContent' => 'lead',
+                        'closeModal'    => 1
+                    )
+                )
+            );
+      	}
+
+      	else {
+      		$leads = $model->getEntities(
```

Comment:
For systems with a large number of leads, this is going to kill the system :-) This will pull every lead and then hydrate it into a Doctrine entity using a ton of memory.  Likely any system with more than 100 leads will crap out.  You'll probably need to add a paginated UI.  May be able to leverage the exiting Views/Lead/list.html.php and some of the IndexAction code.